### PR TITLE
[Navi21] Fixing Batchnorm backward precision issues by adjusting workgroup size (SWDEV-292187, SWDEV-319919)

### PR DESCRIPTION
**Pull request by muralinr**
This fix addresses Batchnorm backward precision issues.

### CODE REVIEW

**Comment by junliume** [comment]
@muralinr has this config been verified by the new change?
> ./bin/MIOpenDriver bnormfp16 -n 256 -c 512 -H 28 -W 28 -m 1 --forw 0 -b 1 -r 1

**Comment by junliume** [comment]
> @muralinr has this config been verified by the new change?
> 
> > ./bin/MIOpenDriver bnormfp16 -n 256 -c 512 -H 28 -W 28 -m 1 --forw 0 -b 1 -r 1

Just tested and seems okay:

>root@ixt-rack-185:/root/MIOpen/build/bin# ./MIOpenDriver bnormfp16 -n 256 -c 512 -H 28 -W 28 -m 1 --forw 0 -b 1 -r 1
MIOpenDriver bnormfp16 -n 256 -c 512 -H 28 -W 28 -m 1 --forw 0 -b 1 -r 1
Backwards prop batch norm verification passed on dx.
Backwards prop batch norm verification passed on dscale.
Backwards prop batch norm verification passed on dbias.
Backwards Prop Batch Norm Verifies on CPU and GPU.

**Review by junliume — DISMISSED** [review]
The rest of the question is why the failure now, and on this particular platform?

**Review by atamazov — CHANGES_REQUESTED** [review]
Please add regression test.

File: src/solver/batchnorm/backward_spatial_single.cpp [hunk 1]
```diff
@@ -131,7 +131,7 @@ BnBwdTrainingSpatialSingle::GetSolution(const ExecutionContext& context,
         else if(in_nhw < (32 * 1024 * 1024) && in_cstride > 512)
         {
             variant    = (n >= 32) ? 1 : 3;
-            xlocalsize = std::min(64 * ((in_cstride + 63) / 64), static_cast<unsigned int>(1024));
```
Comment (atamazov) [inline src/solver/batchnorm/backward_spatial_single.cpp:134]: What is 64 in the old code? Is it connected to the WAVESIZE (which is 32 for Navi21)? Ditto all other places where literal 64 is used.

**Comment by muralinr** [comment]
> The rest of the question is why the failure now, and on this particular platform?

I do see failures when work group size is not power of 2.  When I set it to power of 2, failure tests are passed and performance is improved slightly.  There may be requirement on Navi21 to set work group size to power of 2.

File: src/solver/batchnorm/backward_spatial_single.cpp (hunk 1 above)
Comment (muralinr) [reply src/solver/batchnorm/backward_spatial_single.cpp:134]: yes Artem.  It is related to WAVESIZE (64).

**Comment by muralinr** [comment]
> > @muralinr has this config been verified by the new change?
> > > ./bin/MIOpenDriver bnormfp16 -n 256 -c 512 -H 28 -W 28 -m 1 --forw 0 -b 1 -r 1
> 
> Just tested and seems okay:
> 
> > root@ixt-rack-185:/root/MIOpen/build/bin# ./MIOpenDriver bnormfp16 -n 256 -c 512 -H 28 -W 28 -m 1 --forw 0 -b 1 -r 1
> > MIOpenDriver bnormfp16 -n 256 -c 512 -H 28 -W 28 -m 1 --forw 0 -b 1 -r 1
> > Backwards prop batch norm verification passed on dx.
> > Backwards prop batch norm verification passed on dscale.
> > Backwards prop batch norm verification passed on dbias.
> > Backwards Prop Batch Norm Verifies on CPU and GPU.

All tests are passed for these configs.
MIOpenDriver bnormfp16 -n 256 -c 512 -H 28 -W 28 -m 1 --forw 0 -b 1 -r 1 -i 100 -t 1 -w 0
MIOpenDriver bnormfp16 -n 256 -c 512 -H 18 -W 18 -m 1 --forw 0 -b 1 -r 1 -i 100 -t 1 -w 0
MIOpenDriver bnormfp16 -n 256 -c 512 -H 32 -W 32 -m 1 --forw 0 -b 1 -r 1 -i 100 -t 1 -w 0
MIOpenDriver bnormfp16 -n 256 -c 512 -H 52 -W 52 -m 1 --forw 0 -b 1 -r 1 -i 100 -t 1 -w 0
MIOpenDriver bnormfp16 -n 256 -c 512 -H 7 -W 7 -m 1 --forw 0 -b 1 -r 1 -i 100 -t 1 -w 0
MIOpenDriver bnormfp16 -n 256 -c 512 -H 14 -W 14 -m 1 --forw 0 -b 1 -r 1 -i 100 -t 1 -w 0

**Comment by junliume** [comment]
@muralinr please check CI and it seems some cases are failing:
http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/bn_bwd_spatial_num_fix/2/pipeline

File: src/solver/batchnorm/backward_spatial_single.cpp (hunk 1 above)
Comment (atamazov) [reply src/solver/batchnorm/backward_spatial_single.cpp:134]: Ok. This change looks good for urgent fixing of 5.0. But shall we keep this machinery in a full-blown fix, but switching 64 to 32 for Navi21?

**Comment by atamazov** [comment]
Too many CI failures.

File: src/solver/batchnorm/backward_spatial_single.cpp [hunk 2]
```diff
@@ -144,28 +144,18 @@ BnBwdTrainingSpatialSingle::GetSolution(const ExecutionContext& context,
...
-                variant = 3;
-                xlocalsize =
-                    std::min(64 * ((in_cstride + 63) / 64), static_cast<unsigned int>(1024));
-                xgridsize = c * xlocalsize;
-                ldsgcn    = xlocalsize / 64;
-                ldsnogcn  = xlocalsize;
+                variant    = 3;
+                xlocalsize = 1024;
+                xgridsize  = c * xlocalsize;
+                ldsgcn     = xlocalsize / 64;
```
Comment (zjing14) [inline src/solver/batchnorm/backward_spatial_single.cpp:150]: @muralinr Does 64 here mean WaveSize?

File: src/solver/batchnorm/backward_spatial_single.cpp [hunk 3]
```diff
@@ -144,28 +144,18 @@ BnBwdTrainingSpatialSingle::GetSolution(const ExecutionContext& context,
...
-                else
-                {
-                    xlocalsize = 256;
-                    xgridsize  = 256 * c;
-                }
-                ldsgcn   = xlocalsize / 64;
-                ldsnogcn = xlocalsize;
+                variant    = 0;
+                xlocalsize = 1024;
+                ldsgcn     = xlocalsize / 64;
```
Comment (zjing14) [inline src/solver/batchnorm/backward_spatial_single.cpp:157]: 64 == WaveSize?

File: src/solver/batchnorm/backward_spatial_single.cpp (hunk 2 above)
Comment (junliume) [reply src/solver/batchnorm/backward_spatial_single.cpp:150]: It seems so from comments above: https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1386#discussion_r790980027
CI has failed for some nan issues, we may need to deal with them first.

**Comment by atamazov** [comment]
@zjing14 @junliume @muralinr I've created the branch that detects wavesize and uses it instead of 64 in BN and FusedBN kernels. Going to open PR if it passes tests.

**Comment by atamazov** [comment]
[Informative] Branch name is `bn_bwd_spatial_num_fix_1`

File: src/solver/batchnorm/backward_spatial_single.cpp (hunk 2 above)
Comment (muralinr) [reply src/solver/batchnorm/backward_spatial_single.cpp:150]: > @muralinr Does 64 here mean WaveSize?

yes Jing.

File: src/solver/batchnorm/backward_spatial_single.cpp (hunk 3 above)
Comment (muralinr) [reply src/solver/batchnorm/backward_spatial_single.cpp:157]: yes Jing

**Comment by junliume** [comment]
> [Informative] Branch name is `bn_bwd_spatial_num_fix_1`

CI for the branch:
http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/bn_bwd_spatial_num_fix_1/2/pipeline

**Comment by muralinr** [comment]
> > [Informative] Branch name is `bn_bwd_spatial_num_fix_1`
> 
> CI for the branch: http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/bn_bwd_spatial_num_fix_1/2/pipeline

I tested this branch.  This branch doesn't address an original FP16 issue.  I will provide fix for FP16 only.
root@ixt-rack-185:/testx/MIOpen/build# ./bin/MIOpenDriver bnormfp16 -n 256 -c 512 -H 28 -W 28 -m 1 --forw 0 -b 1 -r 1
MIOpenDriver bnormfp16 -n 256 -c 512 -H 28 -W 28 -m 1 --forw 0 -b 1 -r 1
Backwards prop batch norm verification passed on dx.
Backwards prop batch norm verification failed on dscale: 0.987432
max difference in dscale: 0
Backwards prop batch norm verification passed on dbias.

**Comment by junliume** [comment]
@atamazov I am currently having issues cherry-picking changes made in this PR to 5.0, and reverting #1329. The attempts failed tests for the above configs for unknown reasons.

File: src/solver/batchnorm/backward_spatial_single.cpp [hunk 4]
```diff
@@ -213,6 +205,7 @@ BnBwdTrainingSpatialSingle::GetSolution(const ExecutionContext& context,
             {"MIO_BN_LDS_SIZE", ldsnogcn},
             {"MIO_BN_LDSGCN_SIZE", ldsgcn},
             {"MIO_BN_VARIANT", variant},
+            {"MIO_WAVESIZE", wavesize},
```
Comment (atamazov) [inline src/solver/batchnorm/backward_spatial_single.cpp:208]: [Informative] This is not necessary because BN kernels do not use this symbol.

**Review by atamazov — COMMENTED** [review]
It works somehow.

The previous commit of this PR didn't work on gfx906 when if `lds_reduce2()` is enforced instead of `gcn_reduce2()`. I do not know why the failure disappeared, and this looks suspicious.

File: src/solver/batchnorm/backward_spatial_single.cpp [hunk 5]
```diff
@@ -147,14 +148,15 @@ BnBwdTrainingSpatialSingle::GetSolution(const ExecutionContext& context,
...
-                ldsgcn     = xlocalsize / 64;
+                ldsgcn     = xlocalsize / wavesize;
                 ldsnogcn   = xlocalsize;
             }
             else
             {
                 variant    = 0;
                 xlocalsize = 1024;
-                ldsgcn     = xlocalsize / 64;
+                xgridsize  = 1024 * c;
```
Comment (atamazov) [inline src/solver/batchnorm/backward_spatial_single.cpp:158]: :red_circle: Aha, I see, this is the fix. xgridsize wasn't initialized in previous commit!

**Comment by atamazov** [comment]
@junliume Now I am working on porting this onto 5.0 RC branch.

**Comment by atamazov** [comment]
This PR is still missing the regression test.

**Comment by atamazov** [comment]
The fix for develop should also change "failed" to something like "*FAILED*" or so in the BN driver. We overlooked this failure during initial analysis of the issue, and that led to substantial delays.

**Comment by atamazov** [comment]
Another topic to decide: Do we need more changes from `bn_bwd_spatial_num_fix_1` here.

**Comment by atamazov** [comment]
Let's convert this into draft until all the issues resolved, so we have the full-blown fix
- [ ] Identify the root reason of BN issue.
  - Is it `lds_reduce2()`?
- [ ] Regression test
- [ ] Fix for the "failed" BN driver message
- [ ] :question: More from `bn_bwd_spatial_num_fix_1` ?
- [ ] :question: Is it possible to implement DPP based reduction for Navi21 and avoid LDS usage?

**Comment by atamazov** [comment]
For 5.0 we have #1388.

**Comment by atamazov** [comment]
CI job terminated.

**Comment by junliume** [comment]
@atamazov I think we need this in develop branch too, can we move it back to open? Move [this comment](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1386#issuecomment-1021380312) in a new issue.
@muralinr  regression test needs to be added and then we can have it in develop branch first.

**Review by atamazov — APPROVED** [review]

**Review by junliume — APPROVED** [review]

**Comment by atamazov** [comment]
The following two BN configs are failing with both floats and halves on gfx10 before the fix:
```
-n 256 -c 512 -H 18 -W 18 -m 1 --forw 0 -b 1 -r 1
-n 256 -c 512 -H 28 -W 28 -m 1 --forw 0 -b 1 -r 1
```
These will be added to the regression tests.
